src/range.rs

```diff
@@ -149,16 +149,9 @@ impl<V: Version> Range<V> {
 
     /// Compute the intersection of two sets of versions.
     pub fn intersection(&self, other: &Self) -> Self {
-        Self {
-            segments: Self::intersection_segments(&self.segments, &other.segments),
-        }
-    }
-
-    /// Helper function performing intersection of two interval segments.
-    fn intersection_segments(s1: &[Interval<V>], s2: &[Interval<V>]) -> Vec<Interval<V>> {
-        let mut segments = Vec::with_capacity(s1.len().min(s2.len()));
-        let mut left_iter = s1.iter();
-        let mut right_iter = s2.iter();
+        let mut segments = SmallVec::with_capacity(self.segments.len().min(other.segments.len()));
```

Comment:
Same remark that with `negate`. Though there is a small mistake here. I thought that the number of segments would not go further than the one with the least amount of segments but this is wrong (obvious contradiction with intersection with `Range::any()`). I think the number of segments in the intersection is limited by the sum of the number of segments in each but anyway that is not the most common case.

Important point is that I believe we should initialize with `SmallVec::empty()` instead, and probably extend our smallvec with variants of 3 or 4 elements.